opentech/apply/funds/views.py

```diff
@@ -269,20 +277,34 @@ class UpdateReviewersView(DelegatedViewMixin, UpdateView):
     context_name = 'reviewer_form'
 
     def form_valid(self, form):
-        old_reviewers = set(self.get_object().reviewers.all())
         response = super().form_valid(form)
-        new_reviewers = set(form.instance.reviewers.all())
-
-        added = new_reviewers - old_reviewers
-        removed = old_reviewers - new_reviewers
+        if len(form.cleaned_data.values()) != len(set(form.cleaned_data.values())):  # If any of the users match
+            error_message = _('Users cannot be assigned to multiple roles.')
+            messages.error(self.request, mark_safe(error_message + form.errors.as_ul()))
+            form.add_error(None, error_message)
+            return self.form_invalid(form)
+        added_messages_list = []
+        # Loop through cleaned_data and save reviewers by role type to submission
+        for key, user in form.cleaned_data.items():
```

Comment:
I think this is doing a lot of work and queries for what it is.

If we make the changes to allow the field to include the role object (or the cleaned data) then we do not need to do that again.

`update_or_create` is probably your friend here. That way we don't have to worry about the clean up at the end.

```
object, created = objects.update_or_create(role=role, submission=form.instance, defaults={'reviewer': user})
```